logger/Include/Logger.cpp

```diff
@@ -0,0 +1,260 @@
+#include "Macros.h"
+#include "Logger.h"
+
+//std::unique_ptr<Logger> Logger::instance = nullptr;
+
+Logger::Logger(const LogLevel& level, const std::string& path, const unsigned int amount, const bool do_flush)
+    : queue(DEFAULT_SIZE), local_level{ level }, output_path{ path }, amount{ amount }, end(DEFAULT_END), do_flush(do_flush) {
+    fileInit(this->amount);
+
+    thrd = std::thread([this]() {
+        LogData* msg = nullptr;
+        while (!end || !queue.empty()) {
+            while (queue.pop(msg)) {
+                if (msg) {
+                    flushMessage(*msg);
+                    delete msg;
+                }
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(1));
+        }
+        });
+
+}
+
+Logger& Logger::getInstance(const LogLevel& level, const std::string& path, const unsigned int amount, const bool do_flush) {
+
+    static Logger instance(level, path, amount, do_flush);
+    return instance;
+}
+
+void Logger::fileInit(const unsigned int amount)
+{
+    std::string log_dir{ "Logs" };
+
+    bool error = 0;
+    if (std::filesystem::is_directory(output_path))
+    {
+        log_dir = output_path + "/Logs";
+    }
+    else if (output_path != "")error = 1;
+
+    if (std::filesystem::is_directory(log_dir))
+    {
+        size_t count = std::distance(std::filesystem::directory_iterator{ log_dir }, {});
+        if (count >= amount)
+        {
+            for (const auto& file : std::filesystem::directory_iterator(log_dir))
+            {
+                std::filesystem::remove(file);
+
+                count = std::distance(std::filesystem::directory_iterator{ log_dir }, {});
+                if (count < amount) break;
+            }
+        }
+    }
+    else
+    {
+        std::filesystem::create_directory(log_dir);
+    }
+
+    std::string buff_name = log_dir + "/log_";
+    buff_name += std::format("{:%d-%m-%y-%H_%M_%S}", std::chrono::system_clock::now()) + ".txt";
+
+    file = std::ofstream{ buff_name };
+
+    output_path = buff_name;
+
+    if (error)
+        log("invalid output path, default will be used", "[WARNING]", FUNCTION_NAME, local_level, std::this_thread::get_id());
+}
+
+void Logger::shutDown() {
+    end = true;
+    if (thrd.joinable())
+        thrd.join();
+    file.close();
+}
+
+Logger::~Logger() {
+    shutDown();
+}
+
+void Logger::setOutputPath(const std::string& path)
+{
+
+    std::unique_lock lock{ mutex };
+    output_path = path;
+}
+
+const std::string& Logger::getOutputPath() const
+{
+
+    std::shared_lock lock{ mutex };
+    return output_path;
+}
+
+void Logger::setLevel(LogLevel level)
+{
+    std::unique_lock loc{ mutex };
+    local_level = level;
+}
+
+const LogLevel& Logger::getLevel() const
+{
+    std::shared_lock lock{ mutex };
+    return local_level;
+}
+
+void Logger::setFlush(const bool if_flush) {
```

Comment:
No need to mark parameter const, because original value will not be changed since it's passed by copy. You can mark the function const instead.

```suggestion
void Logger::setFlush(bool if_flush) const {
```